criteria/common/src/org/immutables/criteria/constraints/DebugExpressionVisitor.java

```diff
@@ -0,0 +1,46 @@
+package org.immutables.criteria.constraints;
+
+import java.io.PrintWriter;
+import java.util.Collections;
+import java.util.Objects;
+
+/**
+ * Used to output expression tree as string. Useful for debugging expressions.
+ */
+public class DebugExpressionVisitor<Void> implements ExpressionVisitor<Void> {
+
+  private final PrintWriter writer;
+
+  private int depth;
+
+  public DebugExpressionVisitor(PrintWriter writer) {
```

Comment:
there's an idea to generalize this to `Appendable` but the thing I don't like about `Appendable` is checked IO exception on every append, so probably not a good idea also. Maybe `Writer` could be considered as `StringWriter` can be used. Anyway, this is not really important

Reply:
Do you mean something like this ? 

```java
DebugExpressionVisitor(Writer writer) {
   this.writer = new PrintWriter(writer);
}
```

Reply:
if I have a `Writer` and debug accepts `PrintWriter` I can always write from the outside, so it's ok the way it is. It can be sometimes useful if we can use more general type inside. I think this is too much overthinking from me for just a debug, sorry ))